_overviews/overview/a-taste-of-scala.md

```diff
@@ -0,0 +1,1329 @@
+
+These examples demonstrate some of the commonly-used sequence methods:
+
+```scala
+// a sample list
+val a = List(10, 20, 30, 40, 10)      // List(10, 20, 30, 40, 10)
+
+a.drop(2)                             // List(30, 40, 10)
+a.dropRight(2)                        // List(10, 20, 30)
+a.dropWhile(_ < 25)                   // List(30, 40, 10)
+a.filter(_ < 25)                      // List(10, 20, 10)
+a.find(_ > 20)                        // Some(30)
+a.head                                // 10
+a.headOption                          // Some(10)
+a.last                                // 10
+a.lastOption                          // Some(10)
+a.slice(2,4)                          // List(30, 40)
+a.tail                                // List(20, 30, 40, 10)
+a.take(3)                             // List(10, 20, 30)
+a.takeRight(2)                        // List(40, 10)
+a.takeWhile(_ < 30)                   // List(10, 20)
+
+val a = List(List(1,2), List(3,4))
+a.flatten                             // List(1, 2, 3, 4)
+
+val nums = List("one", "two")
+nums.map(_.toUpperCase)             // List("ONE", "TWO")
+nums.flatMap(_.toUpperCase)         // List('O', 'N', 'E', 'T', 'W', 'O')
+```
+
+These examples show how the “fold” and “reduce” methods are used to sum up every element in a sequence:
+
+```scala
+val firstTen = (1 to 10).toList            // List(1, 2, 3, 4, 5, 6, 7, 8, 9, 10)
+
+firstTen.fold(100)(_ + _)                  // 155
+firstTen.foldLeft(100)(_ + _)              // 155
+firstTen.reduce(_ + _)                     // 55
+firstTen.reduceLeft(_ + _)                 // 55
+```
+
+There are many (many!) more methods available to Scala collections classes, and they’re demonstrated in the Collections sections of this Overview and in the Reference documentation.
+
+
+### Tuples
+
+<!-- TODO: introduce a tuple as an alternative to a class? -->
+
+The Scala *tuple* is a type that lets you easily put a collection of different types in the same container. For example, given this `Person` case class:
+
+```scala
+case class Person(name: String)
+```
+
+This is how you create a tuple that contains an `Int`, a `String`, and a custom `Person` value:
+
+```scala
+val t = (11, "eleven", Person("Eleven"))
+```
+
+You can access the tuple values by number:
+
+```scala
+t._1   // 11
+t._2   // "eleven"
+t._3   // Person("Eleven")
+```
+
+You can also assign the tuple fields to variables, as shown in the REPL:
+
+````
+scala> val (num, string, name) = (11, "eleven", Person("Eleven"))
+val num: Int = 11
+val string: String = eleven
+val good: Boolean = Person(Eleven)
+````
+
+Tuples are nice for those times when you want to put a collection of heterogenous types in a little collection-like structure.
+
+<!--
+TODO: state how tuples can be like an unstructured class, or for when a class is overkill?
+-->
+
+
+
+## Contextual Abstractions
```

Comment:
I think this section is too detailed too, but I’m not sure how we should present the benefits of implicits without going into all these explanations.

What do you think of the following?

---

Under certain circumstances, the Scala compiler can “write” some parts of your programs. For instance, consider a program that sorts a list of addresses by two criterias: city name and then street name:

~~~ scala
val addresses: List[Address] = ...
addresses.sortBy(address => (address.city, address.street))
~~~

The sorting algorithm needs to compare addresses by first comparing their city names, and then also their street names when the city names are the same. However, we don’t need to manually define this ordering relation because the compiler is able to summon it automatically based on an existing ordering relation for comparing `String` values.

---

What do you think?

Reply:
Yes, this is a tough one. I’ll think about this some more, but I do like what you have here. It describes what can happen without getting into the details, and it’s motivational in that it makes me want to click a “Learn more about this feature” link in this section.